### PR TITLE
Initial version of urdf_utilities

### PR DESCRIPTION
**Pull request by vincentrou**
If we agree on this, I can add test.

### CODE REVIEW

**Comment by tfoote** [comment]
I'd strongly recommend reviewing the name and picking a specific scope. You can always have multiple packages. http://wiki.ros.org/ROS/Patterns/Conventions

**Comment by bmagyar** [comment]
Thanks Tully,  I guess we were a bit too quick with that choice.

**Comment by vincentrou** [comment]
What about those names : `urdf_kinematics_extractor`,  `urdf_kinematics_reader` or `urdf_kinematics` ?

**Comment by mathias-luedtke** [comment]
`urdf_kinematics_utils`? or `_helpers`?

**Comment by bmagyar** [comment]
Stepping back to the original name:
* vehicle_kinematics_utilities
* vehicle_kinematics_helpers

**Comment by mathias-luedtke** [comment]
> Stepping back to the original name: [...] vehicle [...]

Acutally, most of the code is not related to vehicles..
I could imagine to add other URDF/kinematic functions for use with other controllers as well.

**Comment by bmagyar** [comment]
Thank you @ipa-mdl , you are right. Having that in mind keeps the door open to such functions in the future too.

* `ros_control_urdf_tools`
* `controller_urdf_tools`
* `controllers_urdf_tools`
* `urdf_controller_tools`
* `controller_urdf_helpers`
* `controller_urdf_utilities`

**Comment by bmagyar** [comment]
@graiola @efernandez feel free to chime in too ;)

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+## Build ##
+###########
+
+include_directories(
+  include ${catkin_INCLUDE_DIRS}
+)
+
+## Declare a C++ library
+add_library(${PROJECT_NAME}
+ src/${PROJECT_NAME}.cpp
```
Comment (efernandez) [inline CMakeLists.txt:40]: indentation

File: include/urdf_utilities/urdf_utilities.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#ifndef URDF_UTILITIES_H
```
Comment (efernandez) [inline include/urdf_utilities/urdf_utilities.h:1]: License ?

File: include/urdf_utilities/urdf_utilities.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#define URDF_UTILITIES_H
```
Comment (efernandez) [inline include/urdf_utilities/urdf_utilities.h:2]: roslint: `URDF_UTILITIES_URDF_UTILITIES_H`

I'd encourage to add a roslint test

File: include/urdf_utilities/urdf_utilities.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+   */
+  bool getTransformVector(const std::string& joint_name,
+                          const std::string& parent_link_name,
+                          urdf::Vector3& transform_vector);
+
+  /**
+   * \brief Get distance between two joints from the URDF
+   * \param first_joint_name Name of the first joint
+   * \param second_joint_name Name of the second joint
+   * \param distance Distance in meter between first and second joint
```
Comment (efernandez) [inline include/urdf_utilities/urdf_utilities.h:30]: doxygen: `\return ...` missed

Comment (efernandez) [reply include/urdf_utilities/urdf_utilities.h:30]: Same below

File: include/urdf_utilities/urdf_utilities.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+   */
+  bool getJointSteeringLimits(const std::string& joint_name,
+                              double& steering_limit);
+private:
+  std::string base_link_;
+
+  boost::shared_ptr<urdf::ModelInterface> model_;
+};
+
+}
```
Comment (efernandez) [inline include/urdf_utilities/urdf_utilities.h:58]: `}  // namespace urdf_utilities`

**Review by efernandez — CHANGES_REQUESTED** [review]
Minor changes... but I'd basically add `roslint`

**Comment by efernandez** [comment]
Re. the naming... maybe `ros_control_urdf_tools` although I think `tools` usually means scripts one can run. Maybe `ros_control_urdf_utils` or `ros_control_urdf_utilities` (I know we should avoid `utils`, but I can't figure out any other name now).

**Comment by davetcoleman** [comment]
``ros_control_urdf_utils`` +1

**Comment by vincentrou** [comment]
I think the first remark of @tfoote was to avoid the use of `utils` in the name which does not explain what is doing this pkg. It is why I propose `urdf_kinematics_extractor` instead.

I do not understand the reference to `ros_control` in the name because for me it is not related to it.

**Comment by davetcoleman** [comment]
Looking at the code just now, seems like ``urdf_geometry_utils`` would be appropriate. or ``urdf_geometry_extractor``, though I like it less.

I think @tfoote's main concern was that it was too generic a name, but adding ``ros_control`` to it certainly narrows its scope

**Comment by tfoote** [comment]
The postfix_utils doesn't really add information. It basically says that this has useful things in this package. All packages should be generally useful so I'd recommend not even including "utils" as a postfix. As it doesn't add information for introspection/reading clarity.

**Comment by bmagyar** [comment]
* `ros_control_urdf_parser` -- although it doesn't really parse it 
The problem with `*_kinematics_extractor` is that this package should collect all shared functions of controllers, maybe not only relevant to kinematics. 
How about these?
* `ros_control_urdf_common`
* `ros_controllers_urdf_common` - yes, one's code doesn't have to be `ros_control`-related to use this and will probably not have `ros_control` dependencies. 
* `controllers_urdf_common`  -- but this may be too general

**Comment by mathias-luedtke** [comment]
:+1: for [`controller_urdf_helpers`](https://github.com/ros-controls/urdf_utilities/pull/1#issuecomment-299139814)

**Comment by vincentrou** [comment]
@bmagyar For me this package just extract distances and angles from an urdf model. So it can be used by any package that needs those information.
I like the use of `geometry` instead of `kinematics` : 
`urdf_geometry_extractor` or
`urdf_geometry_parser`
If there is not enough proposal !

**Comment by efernandez** [comment]
I won't add more options ;)
So I'd just :+1: `urdf_geometry_parser` or maybe just `urdf_geometry` (removing the `_utils` suffix).

**Comment by mathias-luedtke** [comment]
`urdf_geometry_parser` works for me :+1:

**Comment by bmagyar** [comment]
Why `geometry`? I'll pull out the function names here:

```
bool getTransformVector(const std::string& joint_name,
                          const std::string& parent_link_name,
                          urdf::Vector3& transform_vector);
bool getDistanceBetweenJoints(const std::string& first_joint_name,
                                const std::string& second_joint_name,
                                double& distance);
bool getJointRadius(const std::string& joint_name,
                                double& radius);
bool getJointSteeringLimits(const std::string& joint_name,
double& steering_limit);
static bool isCylinder(const boost::shared_ptr<const urdf::Link>& link)
static bool getWheelRadius(const boost::shared_ptr<const urdf::Link>& wheel_link, double& wheel_radius)
``` 
(let's not discuss function design here just yet)


And I expect similar functions to be added to this package from different controllers later on. 
I still think `ros_controllers_urdf_common` or `ros_controllers_urdf_helpers` or `controllers_urdf_helpers` or `controller_urdf_helpers` would be a better choice.

**Comment by vincentrou** [comment]
Why not `geometry` ? Those functions are here to get the geometry of the robot (distances, radius, angles).
I do not see the relation with `controllers` since those functions only needs an urdf model and joint name. In fact, for now it will only be used by controllers.

**Comment by bmagyar** [comment]
Because I think for future extensions a controller-related name gives us more freedom. 
All in all I think we had a general acceptance of `urdf_geometry_parser`, right? I could go with that one too.

Give me a go on it and I'll prepare the repo.

**Comment by bmagyar** [comment]
Thank you all for taking part in this lengthy debate. I have rename the repo and will be closing this PR now.

@vincentrou the floor is all yours for a new PR ;)

**Pull request closed without merging**